source/FeatureRegistry.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "FeatureRegistry.h"
+#include "logging/LoggerFactory.h"
+
+using namespace std;
+using namespace Aws::Iot::DeviceClient;
+using namespace Aws::Iot::DeviceClient::Util;
+using namespace Aws::Iot::DeviceClient::Logging;
+
+constexpr char FeatureRegistry::TAG[];
+
+std::shared_ptr<Feature> FeatureRegistry::get(const std::string &name) const
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    if (features.count(name))
+    {
+        return features.at(name);
+    }
+    LOGM_WARN(TAG, "Feature, %s, not found in current registry", name.c_str());
+    return nullptr;
+}
+
+void FeatureRegistry::add(std::string name, std::shared_ptr<Feature> featureToAdd)
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    if (!features.count(name))
+    {
+        features[name] = featureToAdd;
+    }
+    else
+    {
+        LOGM_WARN(TAG, "Attemped to add %s to Feature Registry despite it already existing", name.c_str());
+    }
+}
+
+void FeatureRegistry::remove(const std::string &name)
```

Comment:
Since we now intend for nullptr to mean the feature is disabled, the meaning of the function name "remove" is no longer accurate. Consider renaming to "disable" or even "setNull"

Reply:
I think that from a users standpoint, it is essentially removed? They will be able to try and get the feature and we return that it is not there. There are no ways that we add more or less things into the registry after main.cpp runs. We simple have placeholder keys (this will be useful for dynamic reload) but for all intents and purposes, it is removed.

Reply:
Yes, but the intended audience for the source code is developers, not users. If you have a class that is a wrapper around a container, its function names and behaviors are expected to be correspond with other examples of wrappers around containers.